File: doors&keys/awnser.go
package main

import (
	"bufio"
	"fmt"
	"os"
)

func canKnightReachPrincess(hallwayMap string) bool {
	collectedKeys := make(map[rune]bool, 3)

	for _, cell := range hallwayMap {
		if cell >= 'a' && cell <= 'z' {
			collectedKeys[cell] = true
		} else {
			requiredKey := cell + ('a' - 'A') 
			if !collectedKeys[requiredKey] {
				return false
			}
		}
	}

	return true
}

func main() {
	in := bufio.NewReader(os.Stdin)
	out := bufio.NewWriter(os.Stdout)
	defer out.Flush()

	var testCaseCount int
	if _, err := fmt.Fscan(in, &testCaseCount); err != nil {
		return
	}

	for i := 0; i < testCaseCount; i++ {
		var hallwayMap string
		fmt.Fscan(in, &hallwayMap)

		if canKnightReachPrincess(hallwayMap) {
			fmt.Fprintln(out, "YES")
		} else {
			fmt.Fprintln(out, "NO")
		}
	}
}

File: doors&keys/awnser.py
import sys
from typing import Set


def can_knight_reach_princess(hallway_map: str) -> bool:
    collected_keys: Set[str] = set()

    for cell in hallway_map:
        if cell.islower():
            collected_keys.add(cell)
        else:
            required_key = cell.lower()
            if required_key not in collected_keys:
                return False

    return True


def main() -> None:
    input_data = sys.stdin.read().strip().split()
    if not input_data:
        return

    test_case_count = int(input_data[0])
    hallway_strings = input_data[1:1 + test_case_count]

    outputs = []
    for hallway_map in hallway_strings:
        if can_knight_reach_princess(hallway_map):
            outputs.append("YES")
        else:
            outputs.append("NO")

    sys.stdout.write("\n".join(outputs))


if __name__ == "__main__":
    main()
